### PR TITLE
feat: add sso pkce

### PR DESCRIPTION
**Pull request by J0**
## What kind of change does this PR introduce?

Adds support for PKCE to the SSO flow

- We don't introduce a foreign key since we can have relay states which use the implicit flow and won't have an associated  flow state. 

How this was tested:
- On staging with Okta as IDP, on an instance with the branch version of GoTrue uploaded

GoTrue-js PR:
- https://github.com/supabase/gotrue-js/pull/707

### CODE REVIEW

**Comment by kangmingtay** [comment]
> We don't introduce a foreign key since we can have relay states which use the implicit flow and won't have an associated flow state

btw, you can have a foreign key on the `auth.relay_states.flow_state_id` with default NULL

**Comment by J0** [comment]
You’re right, let me amend

**Comment by remorses** [comment]
Hi, when using `signInWithSSO` with PKCE with Next.js you get redirected with the `access_token` in the URL hash but the session is not saved automatically and nothing happens

When will this PR land approximately? Thank you for the great work!

**Comment by J0** [comment]
Hey @remorses,

Thanks for the query - I'll be patching the PR and sending it for re-review later today.

**Comment by J0** [comment]
Oh as an update on having a fkey with default null. I looked again and realized that the column is a UUID and can't be set to null so going to go without a foreign key.

Let me know if anyone has other suggestions though.

File: internal/api/samlacs.go [hunk 1]
```diff
@@ -86,6 +87,13 @@ func (a *API) SAMLACS(w http.ResponseWriter, r *http.Request) error {
 			return badRequestError("SAML RelayState comes from another IP address, try logging in again?")
 		}
 
+		if relayState.FlowStateID != nil {
+			flowState, err = models.FindFlowStateByID(a.db, relayState.FlowStateID.String())
+			if err != nil {
+				return err
+			}
+		}
```
Comment (hf) [inline internal/api/samlacs.go:95]: Won't it be already populated in the `relayState` if you associate it?

File: internal/models/sso.go [hunk 2]
```diff
@@ -154,8 +154,9 @@ type SAMLRelayState struct {
 
 	RedirectTo string `db:"redirect_to"`
 
-	CreatedAt time.Time `db:"created_at" json:"-"`
-	UpdatedAt time.Time `db:"updated_at" json:"-"`
+	CreatedAt   time.Time  `db:"created_at" json:"-"`
+	UpdatedAt   time.Time  `db:"updated_at" json:"-"`
+	FlowStateID *uuid.UUID `db:"flow_state_id" json:"flow_state_id,omitempty"`
```
Comment (hf) [inline internal/models/sso.go:159]: If you put in a `FlowState *FlowState` here and the relevant `belongs_to` flags, it Pop should just give you the object, right?

File: migrations/20230818113222_add_flow_state_to_relay_state.up.sql [hunk 3]
```diff
@@ -0,0 +1 @@
+alter table {{ index .Options "Namespace" }}.saml_relay_states add column if not exists flow_state_id uuid references {{ index .Options "Namespace" }}.flow_state(id) default null;
```
Comment (hf) [inline migrations/20230818113222_add_flow_state_to_relay_state.up.sql:1]: Can you please add `on cascade delete`?

File: internal/models/sso.go (hunk 2 above)
Comment (J0) [reply internal/models/sso.go:159]: You're right-  edited to use  `belongs_to` accordingly.

**Review by hf — APPROVED** [review]
Looks OK, make sure you test this on a live environment though!

**Pull request merged**